geospaas_harvesting/crawlers.py

```diff
@@ -54,29 +136,190 @@ def _http_get(cls, url, request_parameters=None):
+
+    def _start_normalizing(self, **kwargs):
+        """Iterate over the DatasetInfo objects obtained from the crawler and
+        normalize the attributes. Normalizing happens in separate threads to
+        parallelize the I/Os.
+        """
+        # Launch thread which checks the size of the failed ingestions
+        # queue and dumps it to disk when necessary
+        failed_queue_thread = threading.Thread(target=self._thread_manage_failed_normalizing)
+        failed_queue_thread.start()
+
+        try:
+            # Launch normalizing threads
+            with concurrent.futures.ThreadPoolExecutor(
+                    max_workers=self.max_threads,
+                    thread_name_prefix=self.__class__.__name__) as executor:
+                futures = []
+                for dataset_info in self.dataset_infos:
+                    futures.append(executor.submit(
+                        self._thread_get_normalized_attributes,
+                        dataset_info,
+                        **kwargs
+                    ))
+        except KeyboardInterrupt:
+            self.logger.info('Normalizing thread received stopping signal')
+            for future in reversed(futures):
+                future.cancel()
+            self.logger.info(
+                'Cancelled future normalizing threads')
+        finally:
+            self.logger.debug("Normalizing threads are done")
+            self._results.put(Stop)
+            self.logger.debug('Stopping failed queue watcher thread')
+            self._failed.put(Stop)
+            failed_queue_thread.join()
+
+            # raise exceptions from threads
+            for future in concurrent.futures.as_completed(futures):
+                exception = future.exception()
+                if exception:
+                    self.logger.error(
+                        "Exception happened during thread",
+                        exc_info=exception)
+
+    def _thread_get_normalized_attributes(self, dataset_info, **kwargs):
+        """
+        Gets the attributes needed to insert a dataset into the
+        database from its URL, and puts a dictionary containing these
+        attributes in the results queue.
+        If an error occurs while retrieving the attributes, the dataset
+        info and the exception are put in the _failed queue for
+        processing by the dedicated thread.
+        This method is meant to be run in a thread.
+        """
+        self.logger.debug("Getting metadata for '%s'", dataset_info.url)
+        try:
+            normalized_attributes = self.crawler.get_normalized_attributes(dataset_info, **kwargs)
+        except Exception as error:  # pylint: disable=broad-except
+            self.logger.error("Could not get metadata for '%s'", dataset_info.url, exc_info=True)
+            self._failed.put((dataset_info, error), block=True)
+        else:
+            self._results.put(NormalizedDatasetInfo(dataset_info.url, normalized_attributes))
```

Comment:
just use datasetInfo to make the crawler independant from the geospaas database structure